src/compiler/checker.ts

```diff
@@ -25170,19 +25176,28 @@ namespace ts {
 
             function isTypePresencePossible(type: Type, propName: __String, assumeTrue: boolean) {
                 const prop = getPropertyOfType(type, propName);
-                if (prop) {
-                    return prop.flags & SymbolFlags.Optional ? true : assumeTrue;
-                }
-                return getApplicableIndexInfoForName(type, propName) ? true : !assumeTrue;
+                return prop ?
+                    !!(prop.flags & SymbolFlags.Optional) || assumeTrue :
+                    !!getApplicableIndexInfoForName(type, propName) || !assumeTrue;
             }
 
-            function narrowByInKeyword(type: Type, name: __String, assumeTrue: boolean) {
-                if (type.flags & TypeFlags.Union
-                    || type.flags & TypeFlags.Object && declaredType !== type && !(declaredType === unknownType && isEmptyAnonymousObjectType(type))
-                    || isThisTypeParameter(type)
-                    || type.flags & TypeFlags.Intersection && every((type as IntersectionType).types, t => t.symbol !== globalThisSymbol)) {
+            function narrowByInKeyword(type: Type, nameType: StringLiteralType | NumberLiteralType | UniqueESSymbolType, assumeTrue: boolean) {
+                const name = getPropertyNameFromType(nameType);
+                const isKnownProperty = someType(type, t => isTypePresencePossible(t, name, /*assumeTrue*/ true));
+                if (isKnownProperty) {
+                    // If the check is for a known property (i.e. a property declared in some constituent of
+                    // the target type), we filter the target type by presence of absence of the property.
                     return filterType(type, t => isTypePresencePossible(t, name, assumeTrue));
                 }
+                if (assumeTrue) {
+                    // If the check is for an unknown property, we intersect the target type with `Record<X, unknown>`,
+                    // where X is the name of the property.
+                    const recordSymbol = getGlobalRecordSymbol();
+                    if (recordSymbol !== unknownSymbol) {
+                        const record = getTypeAliasInstantiation(recordSymbol, [nameType, unknownType]);
+                        return mapType(type, t => t.flags & TypeFlags.NonPrimitive ? record : getIntersectionType([t, record]));

```

Comment:
This makes `object` into `Record<"whatever", unknown>`, but `object & {}` into `object & {} & Record<"whatever", unknown>`. Isn't the difference in behavior a bit odd? Should this just be checking if `object` is assignable to `t`, rather than the type flags?

Reply:
It's basically just a cosmetic check to turn `object & Record<"foo", unknown>` into `Record<"foo", unknown>`, given that the difference rarely if ever matters. I'm inclined to just remove it and leave the `object & Record<"foo, unknown>` alone.